File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int n = prices.length;
        // minprice found so far from 0 to i - 1
        int minPrice = prices[0];

        for(int i = 1; i < n; ++i) {
           int currProfit = prices[i] - minPrice;
           maxProfit = Math.max(currProfit, maxProfit); // max profit
           minPrice = Math.min(minPrice, prices[i]); // buying price
        }

        return maxProfit;
    }
}

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        min_price_so_far = sys.maxint
        profit = 0

        for i in range(len(prices)):
            min_price_so_far = min(prices[i], min_price_so_far)
            profit = max(profit, prices[i] - min_price_so_far)
        
        return profit